module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	fulfilledHeight, err := storage.GetFulfilledHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulfilled height: %w", err)
+	}
+
+	fulfilledHeightsIn, fulfilledHeightsOut := util.UnboundedChannel()
+	fulfilledHeightsIn <- fulfilledHeight
+
+	p := &Pruner{
+		logger:                logger.With().Str("component", "execution_data_pruner").Logger(),
+		storage:               storage,
+		pruneCallback:         func(ctx context.Context) error { return nil },
+		fulfilledHeightsIn:    fulfilledHeightsIn,
+		fulfilledHeightsOut:   fulfilledHeightsOut,
+		thresholdChan:         make(chan uint64),
+		heightRangeTargetChan: make(chan uint64),
+		lastFulfilledHeight:   fulfilledHeight,
+		lastPrunedHeight:      lastPrunedHeight,
+		heightRangeTarget:     defaultHeightRangeTarget,
+		threshold:             defaultThreshold,
+		metrics:               metrics,
+	}
+	p.cm = component.NewComponentManagerBuilder().
+		AddWorker(p.loop).
+		Build()
+	p.Component = p.cm
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p, nil
+}
+
+// NotifyFulfilledHeight notifies the Pruner of the latest fulfilled height.
+func (p *Pruner) NotifyFulfilledHeight(height uint64) {
+	if util.CheckClosed(p.cm.ShutdownSignal()) {
+		return
+	}
+
+	p.fulfilledHeightsIn <- height
+}
+
+// SetHeightRangeTarget updates the Pruner's height range target.
+// This may block for the duration of a pruning operation.
+func (p *Pruner) SetHeightRangeTarget(heightRangeTarget uint64) error {
+	select {
+	case p.heightRangeTargetChan <- heightRangeTarget:
+		return nil
+	case <-p.cm.ShutdownSignal():
+		return component.ErrComponentShutdown
+	}
+}
+
+// SetThreshold update's the Pruner's threshold.
+// This may block for the duration of a pruning operation.
+func (p *Pruner) SetThreshold(threshold uint64) error {
+	select {
+	case p.thresholdChan <- threshold:
+		return nil
+	case <-p.cm.ShutdownSignal():
+		return component.ErrComponentShutdown
+	}
+}
+
+func (p *Pruner) loop(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case h := <-p.fulfilledHeightsOut:
+			height := h.(uint64)
+			if height > p.lastFulfilledHeight {
+				p.lastFulfilledHeight = height
+			}
+			p.checkPrune(ctx)
+		case heightRangeTarget := <-p.heightRangeTargetChan:
+			p.heightRangeTarget = heightRangeTarget
+			p.checkPrune(ctx)
+		case threshold := <-p.thresholdChan:
+			p.threshold = threshold
+			p.checkPrune(ctx)
+		}
+	}
+}
+
+func (p *Pruner) checkPrune(ctx irrecoverable.SignalerContext) {
+	if p.lastFulfilledHeight-p.lastPrunedHeight > p.heightRangeTarget+p.threshold {
+		pruneHeight := p.lastFulfilledHeight - p.heightRangeTarget
+
+		p.logger.Info().Uint64("prune_height", pruneHeight).Msg("pruning storage")
+		start := time.Now()
+
+		if err := p.storage.Prune(pruneHeight); err != nil {
+			ctx.Throw(fmt.Errorf("failed to prune: %w", err))
+		}
+
+		if err := p.pruneCallback(ctx); err != nil {
```

Comment:
what is the use case for prune callback given that we didn't pass cid or height to it.

Reply:
it's useful if the application wants to perform some logic after every pruning call. In particular, we use it to trigger value log GC